.circleci/config.yml

```diff
@@ -175,64 +160,129 @@ jobs:
-             if [ -n "$CIRCLE_TAG" ]; then
-               DOCKER_TAG="$CIRCLE_TAG"
-             elif [ "$CIRCLE_BRANCH" == "master" ]; then
-               DOCKER_TAG="${CIRCLE_SHA1:0:7}"
-             else
-               DOCKER_TAG=`echo $CIRCLE_BRANCH | sed 's,/,_,g'`
-             fi
-             env | sort
-             echo "export DOCKER_TAG=$DOCKER_TAG" >> $BASH_ENV
-             echo "export DOCKER_IMAGE=mongoose-push" >> $BASH_ENV
-
+          name: Mix build release
+          command: mix do certs.dev, distillery.release
+      - save_cache:
+          key: certs-cache
+          paths:
+            - "priv"
+      - run:
+          name : Build tar archive
+          command: tar -czf mongoose_push.tar.gz -C _build/${MIX_ENV}/rel/mongoose_push .
+      - run: *DOCKER_TAGS
       - run:
           name: Login to DockerHub
           command: docker login -u "${DOCKERHUB_USER}" -p "${DOCKERHUB_PASS}"
-
       - run:
           name: Build docker image
-          command: docker build -f Dockerfile.release --build-arg MIX_ENV=prod -t $DOCKER_IMAGE:$DOCKER_TAG .
+          command: docker build -f Dockerfile.release --build-arg MIX_ENV=prod -t $DOCKERHUB_REPOSITORY/$DOCKER_IMAGE:$DOCKER_TAG .
+      - run:
+          name: Push docker image but do not tag it as the latest
+          command: docker push $DOCKERHUB_REPOSITORY/$DOCKER_IMAGE:$DOCKER_TAG
+
+  integration-tests:
+    machine:
+        image: ubuntu-1604:201903-01
+    steps:
+      - checkout
+      - run:
+          name: Install elixir
+          command: |
+              wget https://packages.erlang-solutions.com/erlang-solutions_1.0_all.deb && \
+              sudo dpkg -i erlang-solutions_1.0_all.deb && \
+              sudo apt-get update && \
+              sudo apt-get install -y esl-erlang=1:21.0 && \
+              sudo apt-get install -y elixir=1.8.0-1
+              wget https://github.com/jwilder/dockerize/releases/download/v0.6.1/dockerize-linux-amd64-v0.6.1.tar.gz && \
+              sudo chmod +x dockerize-linux-amd64-v0.6.1.tar.gz && \
+              sudo tar -C /usr/local/bin -xzvf dockerize-linux-amd64-v0.6.1.tar.gz && \
+              sudo chmod +x /usr/local/bin/dockerize && \
+              sudo rm dockerize-linux-amd64-v0.6.1.tar.gz
+
+      - restore_cache:
+          keys:
+              - mix-cache-{{ .Environment.CACHE_VERSION }}-{{ .Branch }}-{{ .Revision }}
+      - restore_cache:
+          keys:
+              - certs-cache
+      - run: *DOCKER_TAGS
+      - run:
+          name: Setup mocks and MPush
+          command: |
+              docker network create integration || true && \
+              docker run -d -p 4000:4000 -p 4001:4001 --name fcm --network integration mongooseim/fcm-mock-server && \
+              docker run -d -p 2197:2197 --name apns --network integration mobify/apns-http2-mock-server && \
+              echo "echo $PWD/app/priv" >> $BASH_ENV && \
+              docker run -d -v /home/circleci/app/priv:/opt/app/priv \
```

Comment:
How about creating a docker-compose file that would setup all this? It would also be usable for local environment and could reside next to integration tests. The tests do/may assume some things about this mpush configuration, so it actually make more sense to have all this within MongoosePush codebase, not CI configuration. I think the only variables are the docker image name / tag, which can be taken from system env anyway (docker-compose does support such case).